libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,405 @@
+		if (!diagnostics.empty())
+			m_nonemptyDiagnostics.insert(sourceUnitName);
+		params["diagnostics"] = move(diagnostics);
+		m_client.notify("textDocument/publishDiagnostics", move(params));
+	}
+}
+
+bool LanguageServer::run()
+{
+	while (m_state != State::ExitRequested && m_state != State::ExitWithoutShutdown && !m_client.closed())
+	{
+		MessageID id;
+		try
+		{
+			optional<Json::Value> const jsonMessage = m_client.receive();
+			if (!jsonMessage)
+				continue;
+
+			if ((*jsonMessage)["method"].isString())
+			{
+				string const methodName = (*jsonMessage)["method"].asString();
+				id = (*jsonMessage)["id"];
+
+				if (auto handler = valueOrDefault(m_handlers, methodName))
+					handler(id, (*jsonMessage)["params"]);
+				else
+					m_client.error(id, ErrorCode::MethodNotFound, "Unknown method " + methodName);
+			}
+			else
+				m_client.error({}, ErrorCode::ParseError, "\"method\" has to be a string.");
+		}
+		catch (exception const& _exception)
+		{
+			m_client.error(id, ErrorCode::InternalError, "Unhandled exception: "s + boost::diagnostic_information(_exception));
+		}
+		catch (...)
+		{
+			m_client.error(id, ErrorCode::InternalError, "Unhandled exception: "s + boost::current_exception_diagnostic_information());
+		}
+	}
+	return m_state == State::ExitRequested;
+}
+
+bool LanguageServer::checkServerInitialized(MessageID _id)
+{
+	if (m_state != State::Initialized)
+	{
+		m_client.error(_id, ErrorCode::ServerNotInitialized, "Server is not properly initialized.");
+		return false;
+	}
+	else
+		return true;
+}
+
+void LanguageServer::handleInitialize(MessageID _id, Json::Value const& _args)
+{
+	if (m_state != State::Started)
+	{
+		m_client.error(_id, ErrorCode::RequestFailed, "Initialize called at the wrong time.");
+		return;
+	}
+	m_state = State::Initialized;
+
+	// The default of FileReader is to use `.`, but the path from where the LSP was started
+	// should not matter.
+	string rootPath("/");
+	if (Json::Value uri = _args["rootUri"])
+	{
+		rootPath = uri.asString();
+		if (!boost::starts_with(rootPath, "file://"))
+		{
+			m_client.error(_id, ErrorCode::InvalidParams, "rootUri only supports file URI scheme.");
+			return;
+		}
+		rootPath = rootPath.substr(7);
+	}
+	else if (Json::Value rootPath = _args["rootPath"])
+		rootPath = rootPath.asString();
+
+	m_fileRepository = FileRepository(boost::filesystem::path(rootPath));
+	if (_args["initializationOptions"].isObject())
+		changeConfiguration(_args["initializationOptions"]);
+
+	Json::Value replyArgs;
+	replyArgs["serverInfo"]["name"] = "solc";
+	replyArgs["serverInfo"]["version"] = string(VersionNumber);
+	replyArgs["capabilities"]["textDocumentSync"]["openClose"] = true;
+	replyArgs["capabilities"]["textDocumentSync"]["change"] = 2; // 0=none, 1=full, 2=incremental
+
+	m_client.reply(_id, move(replyArgs));
+}
+
+
+void LanguageServer::handleWorkspaceDidChangeConfiguration(MessageID _id, Json::Value const& _args)
+{
+	if (!checkServerInitialized(_id))
+		return;
```

Comment:
It's probably not something to change right now but it strikes me that we would have been having an exception you can throw to report an error to the client. We have this big `try` block in `run()`. It could have a `catch` clause that catches the exception and calls `error()` with the message+code and continues. The calls to `checkServerInitialized()` would be much more natural - right now it looks as if it was only checking a condition but it actually immediately reports an error which was unexpected to me.

Reply:
I'm actually not so sure about it. In the compiler this makes total sense since the control-flow is not that important, but here you might want to know if it gets further down where you answer the request or not.

Reply:
That can be conveyed with naming. For example I'd usually name such a function with `validate...()` and by the fact that it does not return anything I'd naturally expect it to handle failed validation by throwing.

And you always have the option to use `error()` manually and not return for more fine-grained control. But for the most common case where you just want to error out and exit I think that exceptions actually make control flow clearer - you can see that the throw will just end the handler however deep you are and you don't have to worry about whether the caller will ignore your return code.